crypto/initthread.c

```diff
@@ -0,0 +1,248 @@
+
+    if (!CRYPTO_THREAD_init_local(&destructor_key.value,
+                                  ossl_init_thread_destructor))
+        return 0;
+
+    return 1;
+}
+
+void cleanup_thread(void)
+{
+    CRYPTO_THREAD_cleanup_local(&destructor_key.value);
+    destructor_key.sane = -1;
+}
+
+void OPENSSL_thread_stop_ex(OPENSSL_CTX *ctx)
+{
+    /*
+     * TODO(3.0). It would be nice if we could figure out a way to do this on
+     * all threads that have used the OPENSSL_CTX when the OPENSSL_CTX is freed.
+     * This is currently not possible due to the use of thread local variables.
+     */
+    ossl_ctx_thread_stop(ctx);
+}
+
+void OPENSSL_thread_stop(void)
+{
+    if (destructor_key.sane != -1)
+        ossl_init_thread_stop(NULL,
+                              ossl_init_get_thread_local(&destructor_key.value,
+                                                         0, 0));
+}
+
+void ossl_ctx_thread_stop(void *arg)
+{
+    if (destructor_key.sane != -1)
+        ossl_init_thread_stop(arg,
+                              ossl_init_get_thread_local(&destructor_key.value,
+                                                         0, 1));
+}
+
+#else
+
+static void *thread_event_ossl_ctx_new(OPENSSL_CTX *libctx)
+{
+    THREAD_EVENT_HANDLER **hands = NULL;
+    CRYPTO_THREAD_LOCAL *tlocal = OPENSSL_zalloc(sizeof(*tlocal));
+
+    if (tlocal == NULL)
+        return NULL;
+
+    if (!CRYPTO_THREAD_init_local(tlocal,  NULL)) {
+        goto err;
+    }
+
+    hands = OPENSSL_zalloc(sizeof(*hands));
+    if (hands == NULL)
+        goto err;
+
+    if (!CRYPTO_THREAD_set_local(tlocal, hands))
+        goto err;
+
+    return tlocal;
+ err:
+    OPENSSL_free(hands);
+    OPENSSL_free(tlocal);
+    return NULL;
+}
+
+static void thread_event_ossl_ctx_free(void *tlocal)
+{
+    OPENSSL_free(tlocal);
+}
+
+static const OPENSSL_CTX_METHOD thread_event_ossl_ctx_method = {
+    thread_event_ossl_ctx_new,
+    thread_event_ossl_ctx_free,
+};
+
+void ossl_ctx_thread_stop(void *arg)
+{
+    THREAD_EVENT_HANDLER **hands;
+    OPENSSL_CTX *ctx = arg;
+    CRYPTO_THREAD_LOCAL *local
+        = openssl_ctx_get_data(ctx, OPENSSL_CTX_THREAD_EVENT_HANDLER_INDEX,
+                               &thread_event_ossl_ctx_method);
+
+    if (local == NULL)
+        return;
+    hands = ossl_init_get_thread_local(local, 0, 0);
+    if (hands != NULL)
+        ossl_init_thread_stop(arg, hands);
+}
+#endif /* FIPS_MODE */
+
+
+static void ossl_init_thread_stop(void *arg, THREAD_EVENT_HANDLER **hands)
+{
+    THREAD_EVENT_HANDLER *curr, *prev = NULL;
+    int freehands = 1;
+
+    /* Can't do much about this */
+    if (hands == NULL)
+        return;
+
+    curr = *hands;
+    while (curr != NULL) {
+        if (arg != NULL && curr->arg != arg) {
+            curr = curr->next;
+            freehands = 0;
+            continue;
+        }
+        curr->handfn(curr->arg);
+        prev = curr;
+        curr = curr->next;
+        if (prev == *hands)
+            *hands = NULL;
```

Comment:
Doesnt this *hands get returned?

Reply:
I refactored this to make things a bit clearer.